frontend/appflowy_flutter/lib/plugins/ai_chat/presentation/message/ai_message_bubble.dart

```diff
@@ -425,6 +456,256 @@ class RegenerateButton extends StatelessWidget {
+  final bool isInHoverBar;
+  final void Function() onTap;
+  final void Function(bool)? onOverrideVisibility;
+
+  @override
+  State<SaveToPageButton> createState() => _SaveToPageButtonState();
+}
+
+class _SaveToPageButtonState extends State<SaveToPageButton> {
+  final popoverController = PopoverController();
+
+  @override
+  Widget build(BuildContext context) {
+    final userWorkspaceBloc = context.read<UserWorkspaceBloc>();
+    final userProfile = userWorkspaceBloc.userProfile;
+    final workspaceId =
+        userWorkspaceBloc.state.currentWorkspace?.workspaceId ?? '';
+
+    return MultiBlocProvider(
+      providers: [
+        BlocProvider(
+          create: (context) => SpaceBloc(
+            userProfile: userProfile,
+            workspaceId: workspaceId,
+          )..add(const SpaceEvent.initial(openFirstPage: false)),
+        ),
+        BlocProvider(
+          create: (context) => ChatSettingsCubit(),
+        ),
+      ],
+      child: BlocSelector<SpaceBloc, SpaceState, ViewPB?>(
+        selector: (state) => state.currentSpace,
+        builder: (context, spaceView) {
+          return AppFlowyPopover(
+            controller: popoverController,
+            triggerActions: PopoverTriggerFlags.none,
+            margin: EdgeInsets.zero,
+            offset: const Offset(8, 0),
+            direction: PopoverDirection.rightWithBottomAligned,
+            constraints: const BoxConstraints.tightFor(width: 300, height: 400),
+            onClose: () {
+              if (spaceView != null) {
+                context.read<ChatSettingsCubit>().refreshSources(spaceView);
+              }
+              widget.onOverrideVisibility?.call(false);
+            },
+            child: FlowyTooltip(
+              message: LocaleKeys.chat_addToPageButton.tr(),
+              child: FlowyIconButton(
+                width: DesktopAIConvoSizes.actionBarIconSize,
+                hoverColor: AFThemeExtension.of(context).lightGreyHover,
+                radius: widget.isInHoverBar
+                    ? DesktopAIConvoSizes.hoverActionBarIconRadius
+                    : DesktopAIConvoSizes.actionBarIconRadius,
+                icon: FlowySvg(
+                  FlowySvgs.ai_add_to_page_s,
+                  color: Theme.of(context).hintColor,
+                  size: const Size.square(16),
+                ),
+                onPressed: () {
+                  final documentId = getOpenedDocumentId();
+                  if (documentId != null) {
+                    onAddToExistingPage(documentId);
+                  } else {
+                    widget.onOverrideVisibility?.call(true);
+                    if (spaceView != null) {
+                      context
+                          .read<ChatSettingsCubit>()
+                          .refreshSources(spaceView);
+                    }
+                    popoverController.show();
+                  }
+                },
+              ),
+            ),
```

Comment:
Moving them into a widget or a widget builder may improve readability.

Reply:
OK done! Moved actions to separate file and extracted widgets from build to widget methods.